### PR TITLE
Exponential retry and thread pool executors for upload

### PR DESCRIPTION
**Pull request by IvanBorislavovDimitrov**

### CODE REVIEW

File: multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/ResilientCloudControllerClient.java [hunk 1]
```diff
@@ -395,8 +397,11 @@ public void updateServiceSyslogDrainUrl(String serviceName, String syslogDrainUr
     }
 
     @Override
-    public CloudPackage asyncUploadApplication(String applicationName, Path file, UploadStatusCallback callback) {
-        return executeWithRetry(() -> delegate.asyncUploadApplication(applicationName, file, callback));
+    public CloudPackage asyncUploadApplication(String applicationName, Path file, UploadStatusCallback callback, Duration uploadTimeout) {
```
Comment (ikasarov) [inline multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/ResilientCloudControllerClient.java:400]: maybe we could just split this in two methods with a more clear way to see we're calling the exponentialRetry ulpoad method

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppAsyncExecution.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+                                                                                            newApplicationDigest);
+    }
+
+    @Override
+    public String getPollingErrorMessage(ProcessContext context) {
+        CloudApplicationExtended appToProcess = context.getVariable(Variables.APP_TO_PROCESS);
+        return MessageFormat.format(Messages.ERROR_OCCURRED_DURING_APPLICATION_UPLOAD_0, appToProcess.getName());
+    }
+
+    MonitorUploadStatusCallback getMonitorUploadStatusCallback(CloudApplication app, Path file, StepLogger stepLogger, String correlationId,
```
Comment (ikasarov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppAsyncExecution.java:229]: does this need to be an internal class?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/FileUploadThreadPoolConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            if (r1 == null && r2 == null) {
+                return 0;
+            } else if (r1 == null) {
+                return -1;
+            } else if (r2 == null) {
+                return 1;
+            }
+            int p1 = ((PriorityFuture<?>) r1).getPriority();
+            int p2 = ((PriorityFuture<?>) r2).getPriority();
+            return Integer.compare(p1, p2);
```
Comment (ikasarov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/FileUploadThreadPoolConfiguration.java:86]: if HIGHEST priority is actually the lowest integer - 1 - wouldn't this method return the reverse of what we expect - compare(HIGHEST, LOWEST) would return -1 as HIGHEST is a lower int

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 4]
```diff
@@ -116,13 +119,15 @@ public ResponseEntity<FileMetadata> uploadFile(MultipartHttpServletRequest reque
...
         try (InputStream in = new BufferedInputStream(multipartFile.getInputStream(), INPUT_STREAM_BUFFER_SIZE)) {
             var startTime = LocalDateTime.now();
-            FileEntry fileEntry = fileService.addFile(ImmutableFileEntry.builder()
-                                                                        .space(spaceGuid)
-                                                                        .namespace(namespace)
-                                                                        .name(multipartFile.getOriginalFilename())
-                                                                        .size(BigInteger.valueOf(multipartFile.getSize()))
-                                                                        .build(),
-                                                      in);
+            FileEntry fileEntry = fileStorageThreadPool.submit(new FileUploadThreadPoolConfiguration.PriorityCallable<>(FileUploadThreadPoolConfiguration.Priority.LOWEST,
```
Comment (ikasarov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:122]: maybe extract the callable construction, looks hard to parse (same with the other call above)

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/FileUploadThreadPoolConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.concurrent.TimeoutException;
+
+import javax.inject.Inject;
+
+import org.cloudfoundry.multiapps.controller.core.util.ApplicationConfiguration;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class FileUploadThreadPoolConfiguration {
```
Comment (ikasarov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/FileUploadThreadPoolConfiguration.java:21]: There are 2 internal classes and 1 enum here, maybe it'll be better if they are outside.

File: multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/util/ResilientCloudOperationExecutor.java [hunk 6]
```diff
@@ -38,6 +47,27 @@ public ResilientCloudOperationExecutor withStatusesToIgnore(HttpStatus... status
...
+        int waitTimeBetweenRetriesInMillis = DEFAULT_TIMEOUT_RETRY_WAIT_TIME_IN_MILLIS;
+        int retryIndex = 1;
+        for (int i = 1; i < RETRY_COUNT_RESPONSE_TIME.size(); i++) {
+            try {
+                return operation.apply(RETRY_COUNT_RESPONSE_TIME.get(retryIndex++));
+            } catch (RuntimeException e) {
+                handle(e);
+                if (e.getCause() instanceof java.util.concurrent.TimeoutException
+                    || e instanceof io.netty.handler.timeout.TimeoutException) {
+                    LOGGER.info("Waiting: {} millis before retrying with timeout of: {} millis", waitTimeBetweenRetriesInMillis,
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/util/ResilientCloudOperationExecutor.java]: done

File: multiapps-controller-process/src/main/java/module-info.java [hunk 7]
```diff
@@ -54,7 +54,7 @@
     requires io.netty.handler;
     requires io.netty.transport;
 
-    requires static java.compiler;
     requires static org.immutables.value;
+    requires com.google.errorprone.annotations;
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/module-info.java]: IDE imports it for some reason, removed

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppAsyncExecution.java [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        context.getStepLogger()
+               .infoWithoutProgressMessage(Messages.TIME_ELAPSED_FOR_UPLOAD_0_IN_MILLIS, timeElapsedForUpload);
+        return processCloudPackage(context, client, cloudPackage);
+    }
+
+    private ApplicationToUploadContext buildApplicationToUploadContext(ProcessContext context,
+                                                                       CloudApplicationExtended applicationToProcess) {
+        MtaArchiveElements mtaArchiveElements = context.getVariable(Variables.MTA_ARCHIVE_ELEMENTS);
+        String moduleFileName = mtaArchiveElements.getModuleFileName(applicationToProcess.getModuleName());
+        String spaceGuid = context.getVariable(Variables.SPACE_GUID);
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppAsyncExecution.java]: done

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppAsyncExecution.java (hunk 2 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppAsyncExecution.java:224]: it was that way before and uses methods from the UploadStep

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/FileUploadThreadPoolConfiguration.java (hunk 3 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/FileUploadThreadPoolConfiguration.java]: added tests

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java (hunk 4 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java]: refactored

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/FileUploadThreadPoolConfiguration.java (hunk 5 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/FileUploadThreadPoolConfiguration.java:18]: refactored as separate classes

File: multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/ResilientCloudControllerClient.java (hunk 1 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/ResilientCloudControllerClient.java]: done

File: multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/util/ResilientCloudOperationExecutor.java [hunk 9]
```diff
@@ -38,6 +48,27 @@ public ResilientCloudOperationExecutor withStatusesToIgnore(HttpStatus... status
...
+        int waitTimeBetweenRetriesInMillis = DEFAULT_TIMEOUT_RETRY_WAIT_TIME_IN_MILLIS;
+        int retryIndex = 1;
+        for (int i = 1; i < RETRY_COUNT_RESPONSE_TIME_BACKOFF.size(); i++) {
+            try {
+                return operation.apply(RETRY_COUNT_RESPONSE_TIME_BACKOFF.get(retryIndex++));
+            } catch (RuntimeException e) {
+                handle(e);
+                if (e.getCause() instanceof java.util.concurrent.TimeoutException
+                    || e instanceof io.netty.handler.timeout.TimeoutException) {
+                    LOGGER.info("Waiting: {} ms before retrying with timeout of: {} ms", waitTimeBetweenRetriesInMillis,
```
Comment (ikasarov) [inline multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/util/ResilientCloudOperationExecutor.java:61]: maybe extract message in an external constant. but either way is ok

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 10]
```diff
@@ -116,13 +120,10 @@ public ResponseEntity<FileMetadata> uploadFile(MultipartHttpServletRequest reque
...
         try (InputStream in = new BufferedInputStream(multipartFile.getInputStream(), INPUT_STREAM_BUFFER_SIZE)) {
             var startTime = LocalDateTime.now();
-            FileEntry fileEntry = fileService.addFile(ImmutableFileEntry.builder()
-                                                                        .space(spaceGuid)
-                                                                        .namespace(namespace)
-                                                                        .name(multipartFile.getOriginalFilename())
-                                                                        .size(BigInteger.valueOf(multipartFile.getSize()))
-                                                                        .build(),
-                                                      in);
+            FileEntry fileEntry = fileStorageThreadPool.submit(new PriorityCallable<>(PriorityFuture.Priority.LOWEST,
```
Comment (ikasarov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:123]: 'new PriorityCallable<>' could also be a separate method for improved readability

File: multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/util/ResilientCloudOperationExecutor.java (hunk 9 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/util/ResilientCloudOperationExecutor.java]: extracted

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java (hunk 10 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java]: done

File: multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/ResilientCloudControllerClient.java [hunk 11]
```diff
@@ -397,10 +397,21 @@ public void updateServiceSyslogDrainUrl(String serviceName, String syslogDrainUr
...
+    public CloudPackage asyncUploadApplicationWithExponentialBackoff(String applicationName, Path file, Duration uploadTimeout) {
+        return executeWithExponentialRetry(timeout -> delegate.asyncUploadApplicationWithExponentialBackoff(applicationName, file,
+                                                                                                            timeout));
+    }
+
+    @Override
+    public CloudPackage asyncUploadApplicationWithExponentialBackoff(String applicationName, Path file, UploadStatusCallback callback,
+                                                                     Duration uploadTimeout) {
+        throw new UnsupportedOperationException("Custom upload timeout must not be used!");
+    }
```
Comment (vkalapov) [inline multiapps-controller-client/src/main/java/org/cloudfoundry/multiapps/controller/client/ResilientCloudControllerClient.java]: This is a bit confusing. If the async upload method in **ResilientCloudControllerClient** shouldn't have a custom upload timeout, why doesn't the method on line 400 also throw an exception, since it has **uploadTimeout** as parameter?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/client/LoggingCloudControllerClient.java [hunk 12]
```diff
@@ -585,16 +585,21 @@ public void updateServiceSyslogDrainUrl(String serviceName, String syslogDrainUr
...
+        return delegate.asyncUploadApplicationWithExponentialBackoff(applicationName, file, uploadTimeout);
     }
 
     @Override
-    public CloudPackage asyncUploadApplication(String applicationName, Path file, UploadStatusCallback callback,
-                                               Duration initialRequestTimeout) {
+    public CloudPackage asyncUploadApplicationWithExponentialBackoff(String applicationName, Path file, UploadStatusCallback callback,
+                                                                     Duration uploadTimeout) {
+        throw new UnsupportedOperationException("Custom upload timeout must not be used!");
+    }
```
Comment (vkalapov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/client/LoggingCloudControllerClient.java]: Same comment as in **ResilientCloudControllerClient** for method in line 588

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/ValidateDeployParametersStep.java [hunk 13]
```diff
@@ -176,17 +187,19 @@ private MergedArchiveStreamCreator getMergedArchiveStreamCreator(List<FileEntry>
...
+                                                                                                                   .space(context.getVariable(Variables.SPACE_GUID))
+                                                                                                                   .namespace(context.getVariable(Variables.MTA_NAMESPACE))
+                                                                                                                   .operationId(context.getExecution()
+                                                                                                                                       .getProcessInstanceId())
+                                                                                                                   .size(size)
+                                                                                                                   .build(),
+                                                                                                 archiveStreamWithName.getArchiveStream())))
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/ValidateDeployParametersStep.java]: done

**Review by vkalapov — APPROVED** [review]

**Review by vkalapov — APPROVED** [review]
